### ISSUE TITLE
bibliography_entries()  - check for duplicated cite keys?

### ISSUE THREAD

**Issue by RLumSK** [mitchelloharawild/vitae#66]
## Background
While using the package I was wondering why some of my publications were never shown in the output until I realised that the biber engine (bibtex engine should be similar) does not 
like duplicated cite keys. 

## Relevance 

If the publications get extracted from ORCID using, e.g., the R package 'rorcid', the cite keys are not necessarily unique. However, if one uses 'vitae' (great package by the way!), one may want to automatise the workflow and do not handle publication records manually. 

## Proposed solution 

The output of `bibliography_entries()` is a tibble, I guess it would be rather easy to replace the cite keys by unique entries using the others fieds (or just assign an index). Of course, it is not a bug, but it would make 'vitae' more powerful. 

 ## Update 1

I just realised that the issue roots deeper: the call `RefManageR::ReadBib(...)` takes care of the duplicated cite key entries, the problem is that the new cite keys do not fit anymore to the original BibTeX file which is called in `knit_print.vitae_bibliography()` and this causes the trouble. 
Maybe it is a solution to proceed with the object created by `RefManageR::ReadBib(...)` but I cannot tell, I never used this package before. 

 ## Update 2

I solved this in a script for my CV via importing the BIB-file via `RefManageR::ReadBib()` into a temporary file and then reimport it via 

```r 
RefManageR::toBiblatex( RefManageR::ReadBib(bib_file, check = FALSE),
 .opts = list(style = "Bibtex"))
```
## Update 3

The proposed solution does not work reliably due to some character conversion problem (https://github.com/ropensci/RefManageR/issues/61).

## Update 4

Last entry to provide an intermediate solution until the underlying problem gets fixed: 

```r
pubs <- rorcid::orcid_citations("<<ORCID>>")
citations <- pubs$citation

pat <- regexpr("@[a-z]+{[^{]*?(?=,.*?[[:ascii:]]+\\s*?=)", citations, perl = TRUE)
regmatches(citations, pat, invert = FALSE) <-
  as.list(paste0(
    regmatches(citations, pat, invert = FALSE),
    vapply(1:length(pat), function (x)
      paste(sample(letters, 3, replace = TRUE), collapse = ""), character(1))
  ))
```

I admit, it is a little bit clumsy, but it fixes the duplicated cite keys by adding three random letters, making a collusion very unlikely. I tested a few cases and now all the entries were shown in the PDF produced by 'vitae'.

## Update 5

The code can be largely simplified as follows: 

```r
pubs <- rorcid::orcid_citations("<<ORCID>>")
citations <- pubs$citation
pat <- regexpr("@[a-z]+{[^{]*?(?=,.*?[[:ascii:]]+\\s*?=)", citations, perl = TRUE)
regmatches(citations, pat, invert = FALSE) <- 
   make.unique(regmatches(citations, pat, invert = FALSE), "_")

```

**Comment by mitchelloharawild** [comment]
Could you provide a MRE of the bib file you're having trouble with?

If it's a case of having duplicated keys in the bib file, I think `bibliography_entries()` could raise a warning if this exists, but being able to filter/manipulate based on the key value could be useful to some people. So while this key identifier exists (which as far as I know should be unique for a correctly formatted bib file), I think it's best to use this as the citation key (rather than using an index).

**Comment by RLumSK** [comment]
Sorry, you are right, I have should have added an example BibTeX.

```bib
@article{Hawking:1966,
author = {Hawking, S W},
title = {{The occurrence of singularities in cosmology. II}},
journal = {Proceedings of the Royal Society of London. Series A. Mathematical and Physical Sciences},
year = {1966},
volume = {295},
number = {1443},
pages = {490--493},
month = dec
}

@article{Hawking:1966,
author = {Hawking, S W},
title = {{The occurrence of singularities in cosmology}},
journal = {Proceedings of the Royal Society of London. Series A. Mathematical and Physical Sciences},
year = {1966},
volume = {294},
number = {1439},
pages = {511--521},
month = oct
}
``` 

Both articles were published in 1966, and they may have the same cite key. I admit it is a little bit odd and should not happen, but during my tests, I extracted publication records from ORCID and, unfortunately, it happened quite a lot. I double-checked where those records come from, and it turns out that they were taken from Crossref. 

In other words, just a warning may be enough, because it would mean (when using ORCID records) that I have to download the records, modify them (manually) before using them within 'vitae'.

**Comment by mitchelloharawild** [comment]
I think this is more of an issue with RefManageR being quiet about making keys unique:
https://github.com/ropensci/RefManageR/blob/0e123f9c33ea9787919799fdcacd4a2f2467c4d1/R/ReadBib.R#L52
If a warning should be shown, I think it should be shown in the RefManageR package.

**Comment by RLumSK** [comment]
I agree and I guess for the moment I cannot come up with a good solution. From where I stand the problem is that 'vitae' first imports the BIB-file via 'RefManageR' to extract information but then falls back to the original BIB-file for creating the bibliography. Unfortunately, and because of the above-mentioned character conversion issue in 'RefManageR' the temporary file solution does not work. 

Bottom line, I see two ways (both are not very elegant): 'vitae' sanitizes the BIB-files or the user has to do it (as shown above). Eventually, the conversion problem in 'RefManageR' gets fixed and then the temporary file solution would work.

**Comment by mitchelloharawild** [comment]
Closing as the package no longer uses RefManageR (due for CRAN archival), and the offending keys are now easily be identified by a message, and accessed via the `id` column:
``` r
library(vitae)
#> 
#> Attaching package: 'vitae'
#> The following object is masked from 'package:stats':
#> 
#>     filter
cat(readLines("~/github/vitae/refs.bib"), sep = "\n")
#> @article{Hawking:1966,
#>   author = {Hawking, S W},
#>   title = {{The occurrence of singularities in cosmology. II}},
#>   journal = {Proceedings of the Royal Society of London. Series A. Mathematical and Physical Sciences},
#>   year = {1966},
#>   volume = {295},
#>   number = {1443},
#>   pages = {490--493},
#>   month = dec
#> }
#> 
#> @article{Hawking:1966,
#>   author = {Hawking, S W},
#>   title = {{The occurrence of singularities in cosmology}},
#>   journal = {Proceedings of the Royal Society of London. Series A. Mathematical and Physical Sciences},
#>   year = {1966},
#>   volume = {294},
#>   number = {1439},
#>   pages = {511--521},
#>   month = oct
#> }
tibble::as_tibble(bibliography_entries("~/github/vitae/refs.bib"))
#> biblio2yaml: duplicate key Hawking:1966
#> # A tibble: 2 x 9
#>   id     type   author  issued title     `container-title`    page  volume issue
#>   <chr>  <chr>  <list>  <list> <chr>     <chr>                <chr> <chr>  <chr>
#> 1 Hawki… artic… <list … <list… "<span c… Proceedings of the … 490-… 295    1443 
#> 2 Hawki… artic… <list … <list… "<span c… Proceedings of the … 511-… 294    1439
```

<sup>Created on 2020-10-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
